test262/per-file/main.js

```diff
@@ -0,0 +1,225 @@
+    generateChildren(resultsNode);
+  };
+}
+
+function generateChildren(node) {
+  // Drop all children
+  for (const child of Array.prototype.slice.call(node.children)) child.remove();
+
+  // Generate new ones!
+  const results = pathInTree.reduce((acc, x) => acc.children[x], tree);
+
+  summaryLabel.textContent = "/ " + pathInTree.join(" / ");
+  summaryStatusLabel.innerHTML = generateStatus(results.aggregatedResults);
+
+  Object.keys(results.children)
+    .sort()
+    .forEach((childName) => {
+      const child = results.children[childName];
+      const isLeaf = child.children === null;
+      const template = isLeaf ? leafTreeNodeTemplate : nonLeafTreeNodeTemplate;
+      const childNode = template.content.children[0].cloneNode(true);
+      node.appendChild(childNode);
+      childNode.querySelector("#name").textContent = childName;
+      childNode.querySelector("#status").innerHTML = generateStatus(
+        child.aggregatedResults
+      );
+      childNode.querySelector("#url").href =
+        "https://github.com/tc39/test262/tree/main/" +
+        pathInTree.join("/") +
+        "/" +
+        childName;
+      if (!isLeaf) {
+        childNode
+          .querySelectorAll(".tree-node-name, .tree-node-image")
+          .forEach((childNode) => {
+            childNode.onclick = function () {
+              pathInTree.push(childName);
+              generateChildren(node);
+            };
+          });
+      }
+    });
+}
+
+function color(name) {
+  return TestResultColors[name] || "black";
+}
+
+function resultAwareSort(names) {
+  const resultOrder = [
+    TestResult.PASSED,
+    TestResult.FAILED,
+    TestResult.SKIPPED,
+    TestResult.PROCESS_ERROR,
+    TestResult.TODO_ERROR,
+    TestResult.METADATA_ERROR,
+    TestResult.HARNESS_ERROR,
+    TestResult.TIMEOUT_ERROR,
+    TestResult.RUNNER_EXCEPTION,
+    TestResult.DURATION,
+  ];
+
+  return names.sort((a, b) => {
+    const aIndex = resultOrder.indexOf(a);
+    const bIndex = resultOrder.indexOf(b);
+    return aIndex - bIndex;
+  });
+}
+
+function generateStatus(aggregatedResults) {
+  const status = Object.keys(aggregatedResults)
+    .sort()
+    .reduce((acc, mode) => {
+      const stats = aggregatedResults[mode];
+      const total = Object.keys(stats).reduce(
+        (acc, name) => acc + stats[name],
+        0
+      );
+      if (total === 0) return acc;
+      acc.push(`<div class="mode-summary-container">
+        <span class="mode-result-text">${mode}</span>
+        <div class="mode-bar-container">
+            ${resultAwareSort(Object.keys(stats))
+              .map(
+                (x) =>
+                  `<div title="${
+                    TestResultLabels[x] +
+                    ": " +
+                    ((100 * stats[x]) / total).toFixed(2) +
+                    "%"
+                  }" style="width: ${
+                    (100 * stats[x]) / total
+                  }%; height: 20px; background-color: ${color(x)}"></div>`
```

Comment:
This, CxBoog, is why the world of frontend development has moved on to frameworks like react & vue with their comfy component languages

(maybe make the body a block statement and introduce some variables to not do all the calculations inside the template string)

Reply:
I much prefer this tbh lol